print-bgp.c

```diff
@@ -2673,47 +2683,56 @@ bgp_open_print(netdissect_options *ndo,
     optslen = GET_U_1(bgp_open_header->bgpo_optlen);
-    ND_PRINT("\n\t  Optional parameters, length: %u", optslen);
+    if (optslen == BGP_OPEN_NON_EXT_OPT_LEN) {
+      opsttype = GET_U_1(bgp_open_header->bgpo_opttype);
+      if (opsttype == BGP_OPEN_NON_EXT_OPT_TYPE_EXTENDED_LENGTH) {
+        optslen = GET_BE_U_2(bgp_open_header->bgpo_optlen_extended);
+        extended_opt_params = 1;
+        open_size += 3;
+        opt_size += 1;
+      }
+    }
+    ND_PRINT("\n\t  Optional parameters%s, length: %u",
+             extended_opt_params ? " (Extended)" : "", optslen);
 
-    opt = dat + BGP_OPEN_SIZE;
-    length -= BGP_OPEN_SIZE;
+    opt = dat + open_size;
+    length -= open_size;
 
     i = 0;
     while (i < optslen) {
-        uint8_t opt_type, opt_len;
+      uint8_t opt_type;
+      uint16_t opt_len;
 
-        ND_TCHECK_LEN(opt + i, BGP_OPT_SIZE);
-        if (length < BGP_OPT_SIZE + i)
-            goto trunc;
-        bgpopt = (const struct bgp_opt *)(opt + i);
-        opt_type = GET_U_1(bgpopt->bgpopt_type);
-        opt_len = GET_U_1(bgpopt->bgpopt_len);
-        if (BGP_OPT_SIZE + i + opt_len > optslen) {
-            ND_PRINT("\n\t     Option %u, length: %u, goes past the end of the options",
-                      opt_type, opt_len);
-            break;
-        }
+      ND_TCHECK_LEN(opt + i, opt_size);
+      if (length < opt_size + i)
+        goto trunc;
+      bgpopt = (const struct bgp_opt *)(opt + i);
+      opt_type = GET_U_1(bgpopt->bgpopt_type);
+      opt_len = extended_opt_params ? GET_BE_U_2(bgpopt->bgpopt_len)
+                                    : GET_U_1(bgpopt->bgpopt_len);
+      if (opt_size + i + opt_len > optslen) {
+        ND_PRINT(
+            "\n\t     Option %u, length: %u, goes past the end of the options",
+            opt_type, opt_len);
+        break;
+      }
 
-        ND_PRINT("\n\t    Option %s (%u), length: %u",
-                  tok2str(bgp_opt_values,"Unknown",opt_type),
-                  opt_type,
-                  opt_len);
+      ND_PRINT("\n\t    Option %s (%u), length: %u",
+               tok2str(bgp_opt_values, "Unknown", opt_type), opt_type, opt_len);
 
-        /* now let's decode the options we know*/
-        switch(opt_type) {
+      /* now let's decode the options we know*/
+      switch (opt_type) {
 
-        case BGP_OPT_CAP:
-            bgp_capabilities_print(ndo, opt + BGP_OPT_SIZE + i,
-                                   opt_len);
-            break;
+      case BGP_OPT_CAP:
+        bgp_capabilities_print(ndo, opt + opt_size + i, opt_len);
+        break;
 
-        case BGP_OPT_AUTH:
-        default:
-               ND_PRINT("\n\t      no decoder for option %u",
-               opt_type);
-               break;
+      case BGP_OPT_AUTH:
+      default:
+        ND_PRINT("\n\t      no decoder for option %u", opt_type);
+        break;
         }
-        i += BGP_OPT_SIZE + opt_len;
+        i += opt_size + opt_len;
```

Comment:
This would be easier to review if it didn't reindent the existing code.

Reply:
I'll reduce this noise.

Reply:
@fenner fixed.